crates/ide-completion/src/completions/extern_crate.rs

```diff
@@ -0,0 +1,96 @@
+//! Completion for extern crates
+
+use ide_db::{FxHashSet, SymbolKind};
+use syntax::{
+    ast::{self, NameRef},
+    AstNode,
+};
+
+use crate::{context::CompletionContext, CompletionItem, CompletionItemKind};
+
+use super::Completions;
+
+pub(crate) fn complete_extern_crate(
+    acc: &mut Completions,
+    ctx: &CompletionContext<'_>,
+    name_ref: Option<&NameRef>,
+) {
+    let imported_extern_crates: FxHashSet<String> = ctx
+        .token
+        .parent_ancestors()
+        .find_map(ast::SourceFile::cast)
+        .map(|src_file| {
+            src_file
+                .syntax()
+                .children()
+                .into_iter()
+                .filter_map(ast::ExternCrate::cast)
+                .filter_map(|node| node.name_ref())
+                .map(|name| name.to_string())
+                .collect()
+        })
+        .unwrap_or_default();
+
+    let current_txt =
+        name_ref.as_ref().map(|name_ref| name_ref.to_string()).unwrap_or(String::new());
+
+    for name in ctx.scope.extern_crates() {
+        if (!current_txt.is_empty() && !name.to_smol_str().starts_with(&current_txt))
+            || imported_extern_crates.contains(name.to_smol_str().as_str())
+        {
+            continue;
+        }
+
+        CompletionItem::new(
+            CompletionItemKind::SymbolKind(SymbolKind::Module),
+            ctx.source_range(),
+            name.to_smol_str(),
+        )
+        .add_to(acc, ctx.db);
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::tests::completion_list_no_kw;
+
+    #[test]
+    fn can_complete_extern_crate() {
+        let case = r#"
+//- /lib.rs crate:other_crate_a
+// nothing here
+//- /other_crate_b.rs crate:other_crate_b
+pub mod good_mod{}
+//- /lib.rs crate:crate_c
+// nothing here
+//- /lib.rs crate:lib deps:other_crate_a,other_crate_b,crate_c extern-prelude:other_crate_a,crate_c
+extern crate crate_c;
+extern crate oth$0
+mod other_mod {}
+"#;
+
+        let completion_list = completion_list_no_kw(case);
+
+        assert_eq!("md other_crate_a\n".to_string(), completion_list);
+    }
+
+    #[test]
+    fn will_not_complete_existing_import() {
+        let case = r#"
+//- /lib.rs crate:other_crate_a
+// nothing here
+//- /lib.rs crate:crate_c
+// nothing here
+//- /lib.rs crate:other_crate_b
+//
+//- /lib.rs crate:lib deps:other_crate_a,other_crate_b,crate_c extern-prelude:other_crate_a,other_crate_b,crate_c
+extern crate other_crate_b;
+extern crate oth$0
+mod other_mod {}
+"#;
+
+        let completion_list = completion_list_no_kw(case);
+
+        assert_eq!("md other_crate_a\n".to_string(), completion_list);
+    }
+}
```

Comment:
Let's create a new module in the `ide-completion/src/tests` folder instead, you can also check how a completion test is structured for this usually by taking a look at the `use_tree` module in there for example.